node/circuits.js

```diff
@@ -0,0 +1,154 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+"use strict";
+
+var States = require('./states');
+var StateMachine = require('./state_machine');
+
+// The each circuit uses the circuits collection as the "nextHandler" for
+// "shouldRequest" to consult.  Peers use this hook to weight peers both by
+// healthy and other factors, but the circuit only needs to know about health
+// before forwarding.
+
+function AlwaysShouldRequestHandler() { }
+
+AlwaysShouldRequestHandler.prototype.shouldRequest = function shouldRequest() {
+    return true;
+};
+
+var alwaysShouldRequestHandler = new AlwaysShouldRequestHandler();
+
+function Circuits(options) {
+    var self = this;
+    self.circuits = {};
+    self.circuitsByServiceName = {};
+    self.stateOptions = {
+        stateMachine: self,
+        nextHandler: alwaysShouldRequestHandler,
+        timers: options.timers,
+        random: options.random,
+        period: options.period,
+        maxErrorRate: options.maxErrorRate,
+        minRequests: options.minRequests,
+        probation: options.probation,
+    };
+    self.shouldRequestOptions = {};
+}
+
+Circuits.prototype.getCircuit = function getCircuit(callerName, serviceName, endpointName) {
+    var self = this;
+    // TODO escape components
+    var circuitName = callerName + '::' + serviceName + '::' + endpointName;
+    var circuit = self.circuits[circuitName];
+    if (!circuit) {
+        circuit = new Circuit();
+        circuit.circuitName = circuitName;
+        circuit.callerName = callerName;
+        circuit.serviceName = serviceName;
+        circuit.endpointName = endpointName;
+        circuit.shouldRequestOptions = self.shouldRequestOptions;
+        circuit.stateOptions = self.stateOptions;
+        self.circuits[circuitName] = circuit;
+    }
+    return circuit;
+};
+
+Circuits.prototype.handleRequest = function handleRequest(req, buildRes, nextHandler) {
+    var self = this;
+    var callerName = req.headers.cn;
+    if (!callerName) {
+        return buildRes().sendError('BadRequest', 'All requests must have the cn (caller service name) header');
+    }
+    var serviceName = req.serviceName;
+    if (!serviceName) {
+        return buildRes().sendError('BadRequest', 'All requests must have a service name');
+    }
+    return req.withArg1(function withArg1(endpointName) {
+        var circuit = self.getCircuit(callerName, serviceName, endpointName);
+        return circuit.handleRequest(req, buildRes, nextHandler);
+    });
+};
+
+// Called upon membership change to collect services that the corresponding
+// egress node is no longer responsible for.
+Circuits.prototype.updateServices = function updateServices(serviceNames) {
```

Comment:
So the API we have now:
```
update :: (names :: List<name>) -> void
```

But we then use names to only ever answer "do we care about this one?" questions.

Can we instead have the api explicate that usage pattern:
```
update :: (owns :: (name) -> Boolean) -> void
```

Where owns is some predicate function?